File: src/stores/garden.ts
import { create } from 'zustand';
import { invoke } from '@tauri-apps/api/core';
import { useVaultStore } from './vault';
import type {
  ActionFilters,
  ActionGrouping,
  ActionItem,
  GardenConfig,
  GardenCounts,
  GardenList,
  InboxItem,
  ProcessTarget,
  ProjectDetail,
  ProjectItem,
  SomedayItem,
  WaitingItem,
} from '@/types/garden';

export interface GardenUIState {
  /** Sidebar Garden section open/closed. */
  sectionOpen: boolean;
  /** Quick-capture Cmd+I modal visibility. */
  captureOpen: boolean;
}

interface GardenState extends GardenUIState {
  // ---- Cached data ----
  counts: GardenCounts;
  inbox: InboxItem[];
  actions: ActionItem[];
  projects: ProjectItem[];
  waiting: WaitingItem[];
  someday: SomedayItem[];
  config: GardenConfig | null;

  // ---- UI prefs ----
  grouping: ActionGrouping;
  filters: ActionFilters;
  hideCompleted: boolean;

  // ---- Navigation ----
  toggleSection: () => void;
  openCapture: () => void;
  closeCapture: () => void;

  // ---- Loaders ----
  refreshCounts: () => Promise<void>;
  refreshAll: () => Promise<void>;
  loadInbox: () => Promise<void>;
  loadActions: () => Promise<void>;
  loadProjects: () => Promise<void>;
  loadWaiting: () => Promise<void>;
  loadSomeday: () => Promise<void>;
  loadConfig: () => Promise<void>;

  // ---- Inbox ----
  capture: (text: string) => Promise<string>;
  updateInbox: (
    id: string,
    updates: { text?: string; page?: string | null; source?: string | null; energy_hint?: string | null },
  ) => Promise<void>;
  deleteInbox: (id: string) => Promise<void>;
  processInbox: (id: string, target: ProcessTarget) => Promise<void>;

  // ---- Actions ----
  addAction: (item: {
    action: string;
    context?: string;
    project?: string | null;
    energy?: string | null;
    duration?: string | null;
    page?: string | null;
  }) => Promise<string>;
  updateAction: (
    id: string,
    updates: {
      action?: string;
      context?: string;
      project?: string | null;
      energy?: string | null;
      duration?: string | null;
      page?: string | null;
    },
  ) => Promise<void>;
  completeAction: (id: string, done: boolean) => Promise<void>;
  deleteAction: (id: string) => Promise<void>;

  // ---- Projects ----
  addProject: (item: {
    title: string;
    outcome?: string;
    deadline?: string | null;
    area?: string | null;
    page?: string | null;
  }) => Promise<string>;
  updateProject: (
    id: string,
    updates: {
      title?: string;
      outcome?: string;
      status?: string;
      deadline?: string | null;
      area?: string | null;
      page?: string | null;
    },
  ) => Promise<void>;
  deleteProject: (id: string) => Promise<void>;
  loadProjectDetail: (id: string) => Promise<ProjectDetail>;

  // ---- Waiting ----
  addWaiting: (item: {
    what: string;
    from?: string;
    since?: string | null;
    project?: string | null;
    page?: string | null;
  }) => Promise<string>;
  updateWaiting: (
    id: string,
    updates: {
      what?: string;
      from?: string;
      since?: string;
      project?: string | null;
      page?: string | null;
    },
  ) => Promise<void>;
  completeWaiting: (id: string, done: boolean) => Promise<void>;
  deleteWaiting: (id: string) => Promise<void>;

  // ---- Someday ----
  addSomeday: (item: {
    text: string;
    area?: string | null;
    page?: string | null;
  }) => Promise<string>;
  updateSomeday: (
    id: string,
    updates: { text?: string; area?: string | null; page?: string | null },
  ) => Promise<void>;
  deleteSomeday: (id: string) => Promise<void>;

  // ---- Config / pages ----
  setConfig: (config: GardenConfig) => Promise<void>;
  bindPage: (list: GardenList, item_id: string, note_path: string | null) => Promise<void>;
  createPage: (
    list: GardenList,
    item_id: string,
    note_path: string,
    title: string,
  ) => Promise<void>;

  setGrouping: (g: ActionGrouping) => void;
  setFilters: (f: Partial<ActionFilters>) => void;
  clearFilters: () => void;
  setHideCompleted: (v: boolean) => void;

  /** Drop everything — called when the vault closes. */
  reset: () => void;
}

const EMPTY_COUNTS: GardenCounts = { inbox: 0, actions: 0, projects: 0, waiting: 0 };

// Wrap each invoke so refresh is opportunistic — backend errors propagate but
// counts/lists keep flowing instead of getting stuck on stale data.
async function safe<T>(p: Promise<T>, fallback: T): Promise<T> {
  try {
    return await p;
  } catch (e) {
    console.warn('garden invoke failed:', e);
    return fallback;
  }
}

export const useGardenStore = create<GardenState>((set, get) => ({
  // Sidebar section starts collapsed so the file tree owns the space until
  // the user opts into Garden. Restored from localStorage below.
  sectionOpen: typeof window !== 'undefined'
    ? window.localStorage.getItem('mycel.garden.sectionOpen') === '1'
    : false,
  captureOpen: false,
  counts: EMPTY_COUNTS,
  inbox: [],
  actions: [],
  projects: [],
  waiting: [],
  someday: [],
  config: null,
  grouping: 'context',
  filters: {},
  hideCompleted: false,

  toggleSection: () => set((s) => {
    const next = !s.sectionOpen;
    if (typeof window !== 'undefined') {
      window.localStorage.setItem('mycel.garden.sectionOpen', next ? '1' : '0');
    }
    return { sectionOpen: next };
  }),
  openCapture: () => set({ captureOpen: true }),
  closeCapture: () => set({ captureOpen: false }),

  refreshCounts: async () => {
    const counts = await safe(invoke<GardenCounts>('garden_counts'), EMPTY_COUNTS);
    set({ counts });
  },

  refreshAll: async () => {
    await Promise.all([
      get().loadInbox(),
      get().loadActions(),
      get().loadProjects(),
      get().loadWaiting(),
      get().loadSomeday(),
      get().loadConfig(),
      get().refreshCounts(),
    ]);
  },

  loadInbox: async () => {
    const inbox = await safe(invoke<InboxItem[]>('garden_inbox_list'), []);
    set({ inbox });
  },
  loadActions: async () => {
    const actions = await safe(invoke<ActionItem[]>('garden_actions_list'), []);
    set({ actions });
  },
  loadProjects: async () => {
    const projects = await safe(invoke<ProjectItem[]>('garden_projects_list'), []);
    set({ projects });
  },
  loadWaiting: async () => {
    const waiting = await safe(invoke<WaitingItem[]>('garden_waiting_list'), []);
    set({ waiting });
  },
  loadSomeday: async () => {
    const someday = await safe(invoke<SomedayItem[]>('garden_someday_list'), []);
    set({ someday });
  },
  loadConfig: async () => {
    const config = await safe(invoke<GardenConfig>('garden_config_get'), null as unknown as GardenConfig);
    if (config) set({ config });
  },

  capture: async (text) => {
    const id = await invoke<string>('garden_inbox_capture', { text });
    await Promise.all([get().loadInbox(), get().refreshCounts()]);
    return id;
  },
  updateInbox: async (id, updates) => {
    await invoke('garden_inbox_update', { id, updates });
    await get().loadInbox();
  },
  deleteInbox: async (id) => {
    await invoke('garden_inbox_delete', { id });
    await Promise.all([get().loadInbox(), get().refreshCounts()]);
  },
  processInbox: async (id, target) => {
    await invoke('garden_inbox_process', { id, target });
    await Promise.all([
      get().loadInbox(),
      get().loadActions(),
      get().loadProjects(),
      get().loadWaiting(),
      get().loadSomeday(),
      get().refreshCounts(),
    ]);
    // Reference creates a new note in the vault — refresh the file tree so
    // it shows up in the sidebar.
    if (target.kind === 'reference') {
      await useVaultStore.getState().refreshTree();
    }
  },

  addAction: async (item) => {
    const id = await invoke<string>('garden_action_add', { item });
    await Promise.all([get().loadActions(), get().refreshCounts()]);
    return id;
  },
  updateAction: async (id, updates) => {
    await invoke('garden_action_update', { id, updates });
    await get().loadActions();
  },
  completeAction: async (id, done) => {
    await invoke('garden_action_complete', { id, done });
    await Promise.all([get().loadActions(), get().refreshCounts()]);
  },
  deleteAction: async (id) => {
    await invoke('garden_action_delete', { id });
    await Promise.all([get().loadActions(), get().refreshCounts()]);
  },

  addProject: async (item) => {
    const id = await invoke<string>('garden_project_add', { item });
    await Promise.all([get().loadProjects(), get().refreshCounts()]);
    return id;
  },
  updateProject: async (id, updates) => {
    await invoke('garden_project_update', { id, updates });
    await Promise.all([get().loadProjects(), get().loadActions(), get().loadWaiting(), get().refreshCounts()]);
  },
  deleteProject: async (id) => {
    await invoke('garden_project_delete', { id });
    await Promise.all([get().loadProjects(), get().refreshCounts()]);
  },
  loadProjectDetail: async (id) => {
    return await invoke<ProjectDetail>('garden_project_detail', { id });
  },

  addWaiting: async (item) => {
    const id = await invoke<string>('garden_waiting_add', { item });
    await Promise.all([get().loadWaiting(), get().refreshCounts()]);
    return id;
  },
  updateWaiting: async (id, updates) => {
    await invoke('garden_waiting_update', { id, updates });
    await get().loadWaiting();
  },
  completeWaiting: async (id, done) => {
    await invoke('garden_waiting_complete', { id, done });
    await Promise.all([get().loadWaiting(), get().refreshCounts()]);
  },
  deleteWaiting: async (id) => {
    await invoke('garden_waiting_delete', { id });
    await Promise.all([get().loadWaiting(), get().refreshCounts()]);
  },

  addSomeday: async (item) => {
    const id = await invoke<string>('garden_someday_add', { item });
    await get().loadSomeday();
    return id;
  },
  updateSomeday: async (id, updates) => {
    await invoke('garden_someday_update', { id, updates });
    await get().loadSomeday();
  },
  deleteSomeday: async (id) => {
    await invoke('garden_someday_delete', { id });
    await get().loadSomeday();
  },

  setConfig: async (config) => {
    await invoke('garden_config_update', { config });
    set({ config });
  },
  bindPage: async (list, item_id, note_path) => {
    await invoke('garden_bind_page', { list, itemId: item_id, notePath: note_path });
    // Refresh whichever list was touched.
    switch (list) {
      case 'inbox': await get().loadInbox(); break;
      case 'actions': await get().loadActions(); break;
      case 'projects': await get().loadProjects(); break;
      case 'waiting': await get().loadWaiting(); break;
      case 'someday': await get().loadSomeday(); break;
    }
  },
  createPage: async (list, item_id, note_path, title) => {
    await invoke('garden_create_page', {
      list,
      itemId: item_id,
      notePath: note_path,
      title,
    });
    await useVaultStore.getState().refreshTree();
    switch (list) {
      case 'inbox': await get().loadInbox(); break;
      case 'actions': await get().loadActions(); break;
      case 'projects': await get().loadProjects(); break;
      case 'waiting': await get().loadWaiting(); break;
      case 'someday': await get().loadSomeday(); break;
    }
  },

  setGrouping: (g) => set({ grouping: g }),
  setFilters: (f) => set((s) => ({ filters: { ...s.filters, ...f } })),
  clearFilters: () => set({ filters: {} }),
  setHideCompleted: (v) => set({ hideCompleted: v }),

  reset: () => set({
    captureOpen: false,
    counts: EMPTY_COUNTS,
    inbox: [],
    actions: [],
    projects: [],
    waiting: [],
    someday: [],
    config: null,
    filters: {},
  }),
}));

File: src-tauri/src/commands/garden.rs
use crate::core::garden::{self as g, GardenConfig, GardenCounts};
use crate::AppState;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tauri::State;

async fn vault_root(state: &State<'_, AppState>) -> Result<PathBuf, String> {
    let guard = state.vault.lock().await;
    guard
        .as_ref()
        .map(|v| v.root.clone())
        .ok_or_else(|| "No vault open".to_string())
}

fn map_err<T>(r: anyhow::Result<T>) -> Result<T, String> {
    r.map_err(|e| e.to_string())
}

// ---------- Inbox ----------

#[tauri::command]
pub async fn garden_inbox_list(
    state: State<'_, AppState>,
) -> Result<Vec<g::InboxItem>, String> {
    let root = vault_root(&state).await?;
    let f = map_err(g::read_inbox(&root))?;
    Ok(f.items)
}

#[tauri::command]
pub async fn garden_inbox_capture(
    text: String,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let root = vault_root(&state).await?;
    map_err(g::capture_inbox(&root, text))
}

#[tauri::command]
pub async fn garden_inbox_update(
    id: String,
    updates: g::InboxUpdate,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::update_inbox(&root, &id, updates))
}

#[tauri::command]
pub async fn garden_inbox_delete(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::delete_inbox(&root, &id))
}

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ProcessTarget {
    NextAction {
        context: Option<String>,
        project: Option<String>,
        energy: Option<String>,
        duration: Option<String>,
    },
    Project {
        outcome: Option<String>,
        first_action: Option<String>,
        action_context: Option<String>,
    },
    WaitingFor {
        from: Option<String>,
        project: Option<String>,
    },
    Someday {
        area: Option<String>,
    },
    Reference {
        note_path: String,
    },
    Trash,
}

#[derive(Debug, Serialize)]
pub struct ProcessResult {
    pub created_id: Option<String>,
    pub created_action_id: Option<String>,
    pub created_note: Option<String>,
}

#[tauri::command]
pub async fn garden_inbox_process(
    id: String,
    target: ProcessTarget,
    state: State<'_, AppState>,
) -> Result<ProcessResult, String> {
    let root = vault_root(&state).await?;
    let inbox = map_err(g::read_inbox(&root))?;
    let item = inbox
        .items
        .iter()
        .find(|i| i.id == id)
        .ok_or_else(|| format!("Inbox item {id} not found"))?
        .clone();

    let mut out = ProcessResult {
        created_id: None,
        created_action_id: None,
        created_note: None,
    };

    match target {
        ProcessTarget::NextAction {
            context,
            project,
            energy,
            duration,
        } => {
            let new_id = map_err(g::add_action(
                &root,
                g::NewAction {
                    action: item.text.clone(),
                    context: context.unwrap_or_else(|| "@везде".into()),
                    project,
                    energy,
                    duration,
                    page: item.page.clone(),
                },
            ))?;
            out.created_id = Some(new_id);
        }
        ProcessTarget::Project {
            outcome,
            first_action,
            action_context,
        } => {
            let title = item.text.clone();
            let proj_id = map_err(g::add_project(
                &root,
                g::NewProject {
                    title: title.clone(),
                    outcome: outcome.unwrap_or_default(),
                    deadline: None,
                    area: None,
                    page: item.page.clone(),
                },
            ))?;
            out.created_id = Some(proj_id);
            if let Some(action_text) = first_action.filter(|s| !s.trim().is_empty()) {
                let aid = map_err(g::add_action(
                    &root,
                    g::NewAction {
                        action: action_text,
                        context: action_context.unwrap_or_else(|| "@везде".into()),
                        project: Some(title),
                        energy: None,
                        duration: None,
                        page: None,
                    },
                ))?;
                out.created_action_id = Some(aid);
            }
        }
        ProcessTarget::WaitingFor { from, project } => {
            let new_id = map_err(g::add_waiting(
                &root,
                g::NewWaiting {
                    what: item.text.clone(),
                    from: from.unwrap_or_default(),
                    since: None,
                    project,
                    page: item.page.clone(),
                },
            ))?;
            out.created_id = Some(new_id);
        }
        ProcessTarget::Someday { area } => {
            let new_id = map_err(g::add_someday(
                &root,
                g::NewSomeday {
                    text: item.text.clone(),
                    area,
                    page: item.page.clone(),
                },
            ))?;
            out.created_id = Some(new_id);
        }
        ProcessTarget::Reference { note_path } => {
            let title = item.text.lines().next().unwrap_or("Note").to_string();
            map_err(g::create_reference_note(&root, &note_path, &title, ""))?;
            out.created_note = Some(note_path);
        }
        ProcessTarget::Trash => {}
    }

    map_err(g::delete_inbox(&root, &id))?;
    Ok(out)
}

// ---------- Actions ----------

#[tauri::command]
pub async fn garden_actions_list(
    state: State<'_, AppState>,
) -> Result<Vec<g::ActionItem>, String> {
    let root = vault_root(&state).await?;
    let f = map_err(g::read_actions(&root))?;
    Ok(f.items)
}

#[tauri::command]
pub async fn garden_action_add(
    item: g::NewAction,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let root = vault_root(&state).await?;
    map_err(g::add_action(&root, item))
}

#[tauri::command]
pub async fn garden_action_update(
    id: String,
    updates: g::ActionUpdate,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::update_action(&root, &id, updates))
}

#[tauri::command]
pub async fn garden_action_complete(
    id: String,
    done: bool,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::complete_action(&root, &id, done))
}

#[tauri::command]
pub async fn garden_action_delete(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::delete_action(&root, &id))
}

#[tauri::command]
pub async fn garden_actions_clear_completed(
    state: State<'_, AppState>,
) -> Result<usize, String> {
    let root = vault_root(&state).await?;
    map_err(g::clear_completed_actions(&root))
}

// ---------- Projects ----------

#[tauri::command]
pub async fn garden_projects_list(
    state: State<'_, AppState>,
) -> Result<Vec<g::ProjectItem>, String> {
    let root = vault_root(&state).await?;
    let f = map_err(g::read_projects(&root))?;
    Ok(f.items)
}

#[tauri::command]
pub async fn garden_project_add(
    item: g::NewProject,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let root = vault_root(&state).await?;
    map_err(g::add_project(&root, item))
}

#[tauri::command]
pub async fn garden_project_update(
    id: String,
    updates: g::ProjectUpdate,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::update_project(&root, &id, updates))
}

#[tauri::command]
pub async fn garden_project_delete(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::delete_project(&root, &id))
}

#[derive(Debug, Serialize)]
pub struct ProjectDetail {
    pub project: g::ProjectItem,
    pub actions: Vec<g::ActionItem>,
    pub waiting: Vec<g::WaitingItem>,
}

#[tauri::command]
pub async fn garden_project_detail(
    id: String,
    state: State<'_, AppState>,
) -> Result<ProjectDetail, String> {
    let root = vault_root(&state).await?;
    let projects = map_err(g::read_projects(&root))?;
    let project = projects
        .items
        .into_iter()
        .find(|p| p.id == id)
        .ok_or_else(|| format!("Project {id} not found"))?;
    let title = project.title.clone();
    let actions = map_err(g::read_actions(&root))?
        .items
        .into_iter()
        .filter(|a| a.project.as_deref() == Some(title.as_str()))
        .collect();
    let waiting = map_err(g::read_waiting(&root))?
        .items
        .into_iter()
        .filter(|w| w.project.as_deref() == Some(title.as_str()))
        .collect();
    Ok(ProjectDetail {
        project,
        actions,
        waiting,
    })
}

// ---------- Waiting For ----------

#[tauri::command]
pub async fn garden_waiting_list(
    state: State<'_, AppState>,
) -> Result<Vec<g::WaitingItem>, String> {
    let root = vault_root(&state).await?;
    let f = map_err(g::read_waiting(&root))?;
    Ok(f.items)
}

#[tauri::command]
pub async fn garden_waiting_add(
    item: g::NewWaiting,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let root = vault_root(&state).await?;
    map_err(g::add_waiting(&root, item))
}

#[tauri::command]
pub async fn garden_waiting_update(
    id: String,
    updates: g::WaitingUpdate,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::update_waiting(&root, &id, updates))
}

#[tauri::command]
pub async fn garden_waiting_complete(
    id: String,
    done: bool,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::complete_waiting(&root, &id, done))
}

#[tauri::command]
pub async fn garden_waiting_delete(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::delete_waiting(&root, &id))
}

#[tauri::command]
pub async fn garden_waiting_clear_completed(
    state: State<'_, AppState>,
) -> Result<usize, String> {
    let root = vault_root(&state).await?;
    map_err(g::clear_completed_waiting(&root))
}

// ---------- Someday ----------

#[tauri::command]
pub async fn garden_someday_list(
    state: State<'_, AppState>,
) -> Result<Vec<g::SomedayItem>, String> {
    let root = vault_root(&state).await?;
    let f = map_err(g::read_someday(&root))?;
    Ok(f.items)
}

#[tauri::command]
pub async fn garden_someday_add(
    item: g::NewSomeday,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let root = vault_root(&state).await?;
    map_err(g::add_someday(&root, item))
}

#[tauri::command]
pub async fn garden_someday_update(
    id: String,
    updates: g::SomedayUpdate,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::update_someday(&root, &id, updates))
}

#[tauri::command]
pub async fn garden_someday_delete(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::delete_someday(&root, &id))
}

// ---------- Config / Counts ----------

#[tauri::command]
pub async fn garden_config_get(
    state: State<'_, AppState>,
) -> Result<GardenConfig, String> {
    let root = vault_root(&state).await?;
    map_err(g::read_config(&root))
}

#[tauri::command]
pub async fn garden_config_update(
    config: GardenConfig,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::write_config(&root, &config))
}

#[tauri::command]
pub async fn garden_counts(state: State<'_, AppState>) -> Result<GardenCounts, String> {
    let root = vault_root(&state).await?;
    map_err(g::counts(&root))
}

// ---------- Page binding ----------

#[derive(Debug, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum GardenList {
    Inbox,
    Actions,
    Projects,
    Waiting,
    Someday,
}

impl GardenList {
    fn key(&self) -> &'static str {
        match self {
            Self::Inbox => "inbox",
            Self::Actions => "actions",
            Self::Projects => "projects",
            Self::Waiting => "waiting",
            Self::Someday => "someday",
        }
    }
}

#[tauri::command]
pub async fn garden_create_page(
    list: GardenList,
    item_id: String,
    note_path: String,
    title: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::create_reference_note(&root, &note_path, &title, ""))?;
    map_err(g::bind_page(&root, list.key(), &item_id, Some(note_path)))
}

#[tauri::command]
pub async fn garden_bind_page(
    list: GardenList,
    item_id: String,
    note_path: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let root = vault_root(&state).await?;
    map_err(g::bind_page(&root, list.key(), &item_id, note_path))
}
